### DISCUSSION TITLE
Help in how to add new entry to datatable or edit data of a selected row

### DISCUSSION THREAD

**Discussion by BearCooder** [shadcn-ui/ui#2639] [Q&A]
Hello guys,
I am building the Datatable from shadcn [documentation](https://ui.shadcn.com/docs/components/data-table).

I replicated this table step by step like in the instruction and it works perfectly. Now I want to add an ADD button to add a new entry to the table. And and EDIT button that enables to edit a selected row in the table,

However when I look at the documentation the links are all dead ends. I got redirected to the Tanstack page from Shadcn.

When I go to the documentation there is only an editable example but I cannot run it. 
I get Unable to run Embedded Project and a blank page.

Does anyone know how can I achieve my goal? Any help is appreciated please. Thank you!

**Comment by lucasfelixs** [comment]
Hi @BearCooder, first of all I suggest you take a look at this [example](https://ui.shadcn.com/examples/tasks) because it is more complete and helped me a lot, so looking at this example I suggest you add the ADD button in the table header, perhaps aligned to the right, and the EDIT button is in the row menu, you can see in this example that they do this, if you look at the row menu, you will see the edit button. To edit, you can add a [dialog](https://ui.shadcn.com/docs/components/dialog). I'm sure this UX is very good. Good luck and any questions you can send here.

**Comment by chirag-9121** [comment]
**1) To ADD new entry**

By referring to the [basic data table](https://ui.shadcn.com/docs/components/data-table#basic-table) ShadCN docs, you must get an idea of how to setup the columns and data table file. Now for adding a button that adds a new entry in the data table you can simply setup a state and newEntryHandler.

Example-

```
import { columns } from "./columns";
import { DataTable } from "./data-table";

export default function Name(){

const [transaction, setTransaction] = useState({
 amount:0, 
 type:""
})
const [data, setData] = useState([]);  // Will contain the list of transactions

// Add new entry handler function
const addDataHandler = (e, data) => {
 // Can make a post request to backend and once successful can update the data state
 .
 .
 setData((prev) => [...prev, data]);
}

return (
 <form onSubmit={(e) => addDataHandler(e, transaction)>
<input onChange={(e) => setTransaction({ ...transaction, amount: e.target.value })}
            value={transaction.amount}>
<input>
<button type="submit">Add</button>
 <DataTable columns={columns} data={data} />

)}
```

The updated data will be added to the data table.


**2) To EDIT the row**

There is also documentation on [row actions](https://ui.shadcn.com/docs/components/data-table#row-actions) on ShadCN site. We can simply access the row data details using `row.original`.
